[2단계 - 장바구니 협업] 빅터(이우승) 미션 제출합니다.

woose28 opened the pull request:
안녕하세요 포코!!

이번 2단계 미션에서는 백엔드 크루들이 구현한 실제 API를 서비스와 연결하는 작업을 했습니다!!
협업하면서 백엔드 크루들과 설계한 API 명세는 [여기서](https://principled-leopon-6bb.notion.site/API-f47afbd32fa043849fbcc08d8d182d28) 확인 가능합니다!

## 구현 사항
### 필수 요구사항
- [x] 로컬 기반의 장바구니를 서버 기반으로 변경
- [x] API 호스트를 변경해 모든 백엔드 크루들의 API를 테스트할 수 있어야 한다.

### 장바구니 추가 구현 사항
- [x] 장바구니 수량과 재고에 따른 기능 구현
  - [x] 상품의 재고가 0인 경우, 체크 박스와 수량 버튼을 비활성화하고 메시지 출력
  - [x] 상품의 수량이 상품의 재고보다 많을 경우, 수량 증가 버튼을 비활성화하고 메시지 출력

## 실행 방법
1. Git clone
```
git clone https://github.com/woose28/react-shopping-cart-prod.git -b step4
```

2. 의존성 패키지 설치
```
npm install
```

3. 실행!
```
npm run start
```

바쁘신 와중에 리뷰를 해주셔서 감사합니다!

woose28 committed:
docs: 기능 요구사항 목록 업데이트

woose28 committed:
refactor: mock data 수정

- quantity -> stock 수정

woose28 committed:
refactor: 상품 정보 api endpoint 변경

- /products -> /api/products
- pagination params 이름 수정
  - _page -> page
  - _limit -> limit

woose28 committed:
refactor: 장바구니 api endpoint 변경

- /shopping-cart -> /api/carts

woose28 committed:
refactor: 장바구니 상품 추가 api 수정

- body 데이터 키 수정(productId -> id)
- mock 반환 값 키 수정(productData -> product)
- 반환 값 변경에 따른 코드 수정(productData -> product)

woose28 committed:
refactor: 장바구니 수량 변경 api 수정

- body 데이터 키 수정(productId -> id)
- api 레벨 전까지(=api 요청 보내기 전까지) 상품 id는 productId를 사용하도록 수정

woose28 committed:
refactor: 로그아웃 시 장바구니 상태 값을 초기화하도록 수정

woose28 committed:
fix: 로그아웃 confirm 시 true 일 때만 HOME으로 이동하도록 수정

woose28 committed:
feat: 상품 목록 화면의 품절된 상품 UI 처리

- 품절된 상품의 경우 장바구니 아이콘 대신 '품절' 텍스트가 보이도록 수정
- useCount의 handleIncrement max 값 조건 수정

woose28 committed:
feat: 장바구니 페이지 품절 상품 처리 기능 구현

- 품절 상품 메시지 추가
- 체크 박스 비활성화
- 수량 변경 버튼 비활성화

woose28 committed:
refactor: 장바구니 페이지 상품 수량이 재고 이상일 때 처리 기능 구현

- 상품 수량이 재고보다 많을 때 메시지 추가
- 수량 증가 버튼 시 경고 창이 발생하도록 구현
- 수량이 재고보다 많고 수량 감소 버튼을 누를 시 확인 창이 발생하도록 구현

woose28 committed:
refactor: 품절 상품 텍스트 및 색상 수정

- 상품 목록 페이지에서 품절된 상품의 텍스트 내용 및 색상 수정

woose28 committed:
feat: 장바구니 상품 추가 mock 함수에 재고 비교 기능 구현

woose28 committed:
refactor: 장바구니 상품 삭제 api url 수정

- productId를 searchParams로 전달하도록 수정

woose28 committed:
refactor: api 스펙에 맞게 body key 값 수정

- 장바구니 상품 추가 api id -> productId
- 장바구니 수량 변경 api id -> productId

woose28 committed:
refactor: 장바구니 페이지 품절된 상품 가격 0원으로 수정

woose28 committed:
refactor: mock handler 에러 메시지 상수화

woose28 committed:
refactor: Route Mapper 함수 반복 사용 제거

woose28 committed:
refactor: 과도한 memoization 제거

- CartProductCard 내부의 memoization 제거

woose28 committed:
refactor: 의미 없는 스타일 제거

- Button.style.jsx 내부 의미 없는 스타일 제거

woose28 committed:
refactor: 상품 목록 페이지 스크롤 이동 조건 수정

- isLoading이 false 일 때만 스크롤을 이동하도록 수정

woose28 committed:
refactor: 인자에 값을 할당하는 코드 제거

- sendCartProductDeleteRequest 함수 내부 인자에 값을 할당하는 코드 제거

woose28 committed:
refactor: cart.api 요청 함수 내부 cart 변수 제거

- 가독성을 위해 필요 없는 변수 할당 제거

woose28 committed:
refactor: user.api 요청 함수 내부 과도한 구조 분해 할당 제거

woose28 committed:
refactor: api 스펙에 맞게 url 주소 수정

woose28 committed:
refactor: api 스펙에 맞게 수정

- product 데이터의 키 수정
  - imageURL -> imageUrl

woose28 committed:
refactor: api 스펙에 맞게 수정

- 장바구니 상품 추가 url 수정
- 장바구니 조회 url 수정
- 장바구니 상품 삭제 로직 수정

woose28 committed:
refactor: axios 에러 핸들링 로직 수정

- data가 undefined 일 때 예외 처리 추가

woose28 committed:
refactor: 가격 천 단위로 끊어서 표시

woose28 committed:
feat: 동적으로 baseURL 바꾸는 기능 구현

woose28 committed:
refactor: 구조 분해 할당 예외 상황 처리

woose28 committed:
chore: prettier 규칙 수정

- importOrder에 mocks 디렉터리 등록

woose28 committed:
refactor: mock handler export 방식 수정

- 각 handler 파일에서 handler 함수들을 배열로 export 하도록 수정

woose28 committed:
refactor: handleCartDispatch 기본 acetionType 제거

woose28 committed:
refactor: reducer의 action에 구조 분해 할당 적용

woose28 committed:
refactor: 회원 삭제 핸들러 함수 예외처리 추가

woose28 committed:
docs: 기능 요구사항 현황 업데이트

pocojang commented on .prettierrc.json:
```diff
@@ -10,11 +10,12 @@
     "^(@|store)(.*|$)",
     "^(@|hooks)(.*|$)",
     "^(@|api)(.*|$)",
+    "^(@|mocks)(.*|$)",
     "^(@|pages)(.*|$)",
     "^(@|components/common)(.*|$)",
     "^(@|components)(.*|$)",
     "^(@|constants)(.*|$)",
     "^(@|styles)(.*|$)"
```
오호
`@trivago/prettier-plugin-sort-imports` 설정이군요?

woose28 replied:
맞습니다ㅎㅎ 해당 플러그인을 사용하면 import 구문을 정해진 순서에 맞게 정리해 준다고 하여 사용해 봤습니다!
해당 플로그인을 사용하면 import 구문의 순서 일일이 고치지 않아도 돼서 편리했습니다😃

pocojang commented on src/api/cart.api.js:
```diff
@@ -4,46 +4,41 @@ import customInstance from 'api/customInstance';
 
-  const cart = response.data;
-
   alert(ALERT_MESSAGES.PRODUCT_ADDED(quantity));
```
이 코드가 api 레이어에 위치하기에는 좀 애매하지 않을까요?

woose28 replied:
다시 확인하니 alert를 호출하는 코드가 api 레이어에 있으니 부적절해 보이네요..!

alert 함수는 추후에 Snackbar로 대체될 수 있다고 생각합니다. 따라서 UI 레이어에서 alert 함수를 호출하는 게 적절하다고 판단하였고 이를 수정 사항에 반영했습니다.

[관련 커밋 링크 - c99c0c1](https://github.com/woowacourse/react-shopping-cart-prod/pull/71/commits/c99c0c10c6747db90ba351569814d9e55e59688a)

추가적으로 언급해 주신 alert 함수 호출 이외에도 다른 alert 함수 호출도 모두 UI 레이어로 이동시켰습니다.

[관련 커밋 링크 - d1e6038](https://github.com/woowacourse/react-shopping-cart-prod/pull/71/commits/d1e60380dba76f8c1f75a9204838d40f01ba18ca)

pocojang commented on src/api/cart.api.js:
```diff
@@ -4,46 +4,41 @@ import customInstance from 'api/customInstance';
 export const sendDeleteCartProductRequest = async (productIdArray) => {
-  const response = await productIdArray.reduce(sendCartProductDeleteRequest, null);
-  const cart = response.data;
+  await Promise.all(productIdArray.map(deletingRequestMapper));
```
실제 유저와 사용되는 리얼 API에서도 N회 통신하게 되나요?

woose28 replied:
맞습니다..! API를 설계할 때 상품 여러 개를 한 번에 삭제하는 경우를 고려하지 못했습니다😓
해당 부분은 협의가 된 API 명세가 있어 지금 당장 수정은 힘들 거 같습니다🙏
앞으로는 위 코드처럼 동일한 API를 반복적으로 호출하여 낭비되는 경우가 없도록 API 협의하겠습니다!

pocojang commented on src/api/constants.js:
```diff
@@ -1,15 +1,30 @@
+    name: '파랑',
+    url: 'http://13.124.20.12:8080',
+  },
+];
```
`env`에서 일반적인 상수로 바꾸게 되신 계기가 있으실까요?!

woose28 replied:
실제 API 연결할 때 URL을 env 파일로 관리해야 한다는 생각을 미처 하지 못했습니다..!
해당 부분은 수정했습니다!

[관련 커밋 링크 - 1923910](https://github.com/woowacourse/react-shopping-cart-prod/pull/71/commits/19239107ae4f39ccbae6686190ce422630cf3dc9)

pocojang commented on src/api/customInstance.js:
```diff
@@ -16,16 +22,21 @@ if (token !== undefined && nickname !== undefined) {
 
 customInstance.interceptors.response.use((response) => response, handleAPIError);
 
+export const getBaseUrl = () => {};
```
👀

woose28 replied:
API URL을 UI 적으로 바꾸는 기능을 설계할 때 get과 set 함수를 모두 이용하도록 생각했었습니다. 하지만 구현하면서 get 함수는 사용하지 않게 됐지만 함수를 제거하는 걸 깜빡했습니다😓

해당 함수는 제거했고, 관련 커밋은 아래와 같습니다!

[관련 커밋 링크 - 0ee3775](https://github.com/woowacourse/react-shopping-cart-prod/pull/71/commits/0ee3775eb5861d8b1d52577192168b97bc0a1b33)

pocojang commented on src/api/customInstance.js:
```diff
@@ -16,16 +22,21 @@ if (token !== undefined && nickname !== undefined) {
 
+export const getBaseUrl = () => {};
+
+export const setBaseUrl = (index) => {
```
`index`를 받는게 아니라 완성된 `url`을 받는게 낫지 않을까요?

woose28 replied:
제안해 주신 대로 url을 받아서 저장하는 게 조금 더 set의 느낌이 나는 거 같습니다!!

하지만, 해당 함수도 사용이 되지 않는 함수로 확인돼서 제거를 했습니다..!

[관련 커밋 링크 - 4617e4a](https://github.com/woowacourse/react-shopping-cart-prod/pull/71/commits/4617e4aca84062ca222bdeec4d50c7ca2c8b2375)

pocojang commented on src/components/cart/CartProductCard/CartProductCard.jsx:
```diff
@@ -21,28 +26,41 @@ function CartProductCard({
+    decrementCartProduct(productId, currentQuantity);
+  };
+
+  const handleProductDelete = () => deleteProduct([productId]);
```
`productDelete` or `deleteProduct` 통일하는게 어떨까요?

woose28 replied:
deleteProduct로 통일하였습니다. 추가적으로 다른 handler 함수들의 이름들도 `handle+동사+명사` 구조로 수정했습니다!

[관련 커밋 링크 - 91e803f](https://github.com/woowacourse/react-shopping-cart-prod/pull/71/commits/91e803f5a19edba5862eb5ebfea19f3432e1f7c5)

pocojang commented on src/components/cart/CartProductCard/CartProductCard.jsx:
```diff
@@ -51,11 +69,19 @@ function CartProductCard({
+            ? '품절된 상품입니다.'
+            : isMoreThanStock
+            ? '구매 가능한 최대 수량입니다.'
+            : ''}
```
애매한 상황이라 객체로 관리해도 될 것 같은데요
`''`는 빈 문자열을 렌더링해도 되는 경우인가요?

woose28 replied:
해당 부분은 상태 메시지를 객체로 관리해서 렌더링하는 방식으로 수정했습니다!

> ''는 빈 문자열을 렌더링해도 되는 경우인가요?

보여주는 메시지가 없어도 `StatusMessage`컴포넌트(메시지를 보여주는 컴포넌트)의 공간을 확보하기 위해서 빈 문자열을 사용했습니다. 하지만 굳이 사용되지 않은 컴포넌트를 렌더링 할 필요는 없을 거 같네요! 따라서 조건에 따라 `StatusMessage` 컴포넌트를 렌더링하도록 수정했습니다~!


[관련 커밋 링크 - 2d2c683](https://github.com/woowacourse/react-shopping-cart-prod/pull/71/commits/2d2c683f021052f483436c64402c0195789c76f0)

pocojang commented on src/components/common/CheckBox/CheckBox.jsx:
```diff
@@ -3,10 +3,16 @@ import React from 'react';
+  const handleOnClick = () => {
+    if (!disabled) {
+      onClick();
+    }
```
`disabled` 속성만으로는 제어가 어려우셨을까요?!

woose28 replied:
CheckBox의 태그를 button으로 변경하여 disabled 속성을 활용하는 방향으로 수정했습니다~!

[관련 커밋 링크 - 65b2997](https://github.com/woowacourse/react-shopping-cart-prod/pull/71/commits/65b2997842d42d5312fce8aac05bf096e49cb6a7)

pocojang commented on src/components/common/Image/Image.jsx:
```diff
@@ -3,13 +3,13 @@ import React from 'react';
 
-function Image({ src, alt, width = '100%' }) {
+function Image({ src, alt, width = '100%', backgroundColor = '', }) {
   const handleError = (e) => {
     e.target.src = errorImage;
   };
```
👍

pocojang commented on src/components/common/Image/Image.jsx:
```diff
@@ -3,13 +3,13 @@ import React from 'react';
 import errorImage from 'components/common/Image/error.svg';
 
-function Image({ src, alt, width = '100%' }) {
+function Image({ src, alt, width = '100%', backgroundColor = '', }) {
```
`backgroundColor` 없을때의 처리가 좋네요
하지만 단순하게 빈문자열만 넣는 것도 의도되신걸까요?

woose28 replied:
> 하지만 단순하게 빈문자열만 넣는 것도 의도되신걸까요?

의도된 게 맞습니다! 하지만 코멘트를 보고 다시 생각해 보니 backgroundColor 인자의 기본 값으로 빈 문자열이 들어가는 게 자연스러워 보이진 않네요. 따라서 해당 부분은 skeleton view 색상을 기본 값으로 설정하는 방향으로 수정했습니다!

[관련 커밋 링크 - 007c469](https://github.com/woowacourse/react-shopping-cart-prod/pull/71/commits/007c469dd6e5b40411cd6eb43ea083e73d18dc87)

pocojang commented on src/mocks/constant.js:
```diff
@@ -0,0 +1,5 @@
+  NOT_EXIST_PRODUCT: '존재하지 않은 상품입니다.',
+  NOT_EXIST_IN_SHOPPING_CART: '장바구니가 비었거나 장바구니에 존재하지 않는 상품입니다.',
+  EXCEED_STORABLE_QUANTITY: (currentStock, currentQuantity) => `현재 장바구니에 ${currentQuantity}개의 상품이 담겨 있습니다. 새로 추가할 수 있는 최대 수량의 상품은 ${Math.max(currentStock - currentQuantity, 0)}개 입니다.`,
+}
```
👀

woose28 replied:
아이구 파일 끝에 개행이 없었네요..! 해당 부분 수정했습니다!

[관련 커밋 링크 - 1e46181](https://github.com/woowacourse/react-shopping-cart-prod/pull/71/commits/1e46181e38f4fdce79167eda96fcecbe0bc5f0e4)

pocojang commented on src/store/actions/cart.action.js:
```diff
@@ -7,18 +7,13 @@ import {
+    const data = await func(...params);
 
-    dispatch({ type: actionType, payload: { cart } });
+    dispatch({ type: actionType, payload: data && { cart: data.cart } });
```
아무리 `Thunk`를 사용한다지만 잠재적으로 위험한 로직으로 보이는데 문제가 없으신가요?

woose28 replied:
아직까지는 문제가 발생하는 상황을 만나지 못했습니다..!

> 잠재적으로 위험한 로직으로 보이는데

사실 잠재적으로 어떤 위험이 있는지 명확하게 인지하지 못했습니다😓
다만 API 요청 후 반환받은 data가 있지만 data에 cart key로 값이 없다면 `payload: data && { cart: data.cart }` 코드에서 문제가 발생하지 않을까? 생각이 듭니다.
즉, 통신 후 반환받은 값이 제대로 paylaod에 담기지 않는 문제가 발생할 것으로 추측됩니다.

따라서 아래처럼 반환받은 data를 그대로 payload에 담는 방식으로 수정하려 하는데 적절한 수정 방향일까요?

```javascript
const data = await func(...params);

dispatch({ type: actionType, payload: data });
```

pocojang replied:
[Redux 유틸](https://github.com/redux-utilities)과 같은 코드를 구성하고 싶다면 타 라이브러리를 참고해보세요!

[redux-actions](https://github.com/redux-utilities/redux-actions)은 과거 정말 많이 사용되던 헬퍼인데요
@woose28 가 구성한 코드를 보면 헬퍼에 가깝다고 보여요.
그러다보니 이게 카트 액션 구성에 포함되어야하는지도 잘 모르겠고요!

pocojang commented on src/store/actions/cart.action.js:
```diff
@@ -7,18 +7,13 @@ import {
-    dispatch({ type: actionType, payload: { cart } });
+    dispatch({ type: actionType, payload: data && { cart: data.cart } });
   } catch ({ message }) {
     alert(message);
```
마찬가지입니다
`Thunk`를 사용하지만 `Dispatch`에 어울리지 않는 코드로 판단되네요

woose28 replied:
동의합니다! dispatch라고 명명한 함수 내부에서 alert 함수를 호출하는 것은 자연스러워 보이지 않네요! 해당 부분은 [위 피드백](https://github.com/woowacourse/react-shopping-cart-prod/pull/71#discussion_r895086380)을 반영하면서 수정했습니다!

pocojang commented on src/App.jsx:
```diff
@@ -9,25 +9,22 @@ import { COMMON_PAGES, NON_USER_PAGES, USER_PAGES } from 'pages';
 import ProtectedRoute from 'components/route/ProtectedRoute';
 import PageTemplate from 'components/template/PageTemplate/PageTemplate';
 
+const routeComponentMapper = ({ path, element }) => (
+  <Route key={path} path={path} element={element} />
+);
```
`useRoutes()` 대체가능하지 않을까요~

https://stackblitz.com/github/remix-run/react-router/tree/main/examples/route-objects?file=src/App.tsx

woose28 replied:
페어와 함께 작업할 때도 `useRoutes`의 사용을 고려했었는데요. 현재 Router 구조에서 `useRoutes`를 사용하는 게 오히려 가독성이 떨어진다고 판단하여 `useRoutes`를 사용하지 않았습니다. 또한 `PageTemplate`이라는 컴포넌트를 Route로 감싸서 Routes의 바로 아래에 위치 시키고 있기 때문에 `useRoutes`를 사용하게 된다면 Routes -> Route -> Routes 형태로 중첩이 돼서 에러가 발생합니다. 따라서 `useRoutes`를 적용시키는 것은 어려울 거 같습니다..!

현재 프로젝트에서 `useRoutes` 적용시키진 못했지만, Route 구조가 조금 더 간단한 프로젝트에서는 해당 hook을 이용해 보도록 하겠습니다!!

pocojang replied:
이미 많은 고민을 해보셨군요 👍 
잘하셨습니다!

pocojang reviewed (changes requested):
@woose28 👋 
대타로 왔습니다

## 역할 분리

디렉터리 분리는 되어있으나 코드 간에 역할이 중첩되거나 애매하게 나뉘어져있는 경우가 종종 보입니다.

## 비동기 액션

`Redux`를 사용하시며 비동기 액션을 다루는 점은 좋지만 그만큼 비대해지는 경우가 있습니다
물론 `Thunk`를 사용한다지만 그래도 이런 결과를? 이런 내용을? `dispatch` & `action`에 담아야할까? 고민이 드는 경우가 종종 보여요

`store`에서는 `react`를 모르게 `react`에서도 `redux`를 모르도록 분리해보는 것도 좋습니다 (Hooks로 연결하라는 의미는 아닙니다)

## 사용성

버그가 좀 있는데요!
좀 더 꼼꼼하게 확인해보세요~
제가 일일이 확인해보는 것보다는 무엇이 안되는지도 직접 확인해보시면 좋겠습니다

이외에는 코멘트 리뷰를 확인해보세요 👍

woose28 committed:
refactor: 장바구니 상품 추가 성공 시 호출되는 alert 호출 위치 수정

- api 레이어에서 alert 호출이 적절하지 않다고 판단
- 추후 Snackbar 구현을 고려하여 alert 호출 위치를 UI 레이어로 이동

woose28 committed:
refactor: api 에러 발생 시 호출되는 alert 위치 수정

- alert 호출은 UI 레이어에서 발생되는 게 적절하다고 판단
- 따라서 alert 호출 위치를 UI 레이어로 이동

woose28 committed:
refactor: API URL env 파일로 이동

- API URL를 코드 상에서 노출시키지 않기 위해서 env 파일에 API URL을 선언하여 사용하도록 수정

woose28 committed:
refactor: 사용 안 되는 getBaseUrl 함수 제거

woose28 committed:
refactor: 사용 안 되는 setBaseUrl 함수 제거

woose28 committed:
refactor: session storage에 저장된 API URL 불러오는 코드 수정

- 가독성을 향상시키기 위해 구조 분해 할당을 활용하여 코드 라인 수를 감소

woose28 committed:
refactor: 이벤트 handler 함수 이름 수정

- 통일성을 위해 handler 함수 이름은 handle+동사+명사 구조로 명명

woose28 committed:
refactor: 장바구니 상품 카드의 상태 메시지 로직 변경

- 가독성 향상을 위해 상태 메시지를 객체로 관리
- 렌더링 되는 요소를 줄이기 위해서 상태 메시지 렌더링 조건을 수정

woose28 committed:
refactor: CheckBox 컴포넌트 html 태그 button으로 수정

- disabled 속성을 활용하기 위해서 CheckBox의 태그를 div 태그에서 button 태그로 수정

woose28 committed:
refactor: Image 컴포넌트 backgroundColor 인자 기본 값 변경

- backgroundColor 인자의 의미에 맞게 기본 값을 빈 문자열에서 skeleton 색상으로 변경

woose28 committed:
style: 필요 없는 주석 제거

woose28 committed:
refactor: 파일 끝 개행 추가

pocojang commented:
@woose28 
리뷰 반영 확인했습니다
~6/19일까지는 PR 보존해놓을테니 반영하실 것이 있다면 반영하셔도 괜찮습니다

pocojang reviewed (approved):
@woose28
이제 시간이 지나 머지합니다.
리뷰 반영 잘 확인했어요 👍 

## 역할 분리

특정 로직들의 역할 분리가 애매한 부분이 계속 보였고 그에 대한 피드백을 종종 나눴는데요.
지금의 장바구니정도의 볼륨이라면 문제가 없지만 향후 스케일 업된다면 확장 가능성을 잃을 수 있고 횡단 관심사에 대한 처리가 안될 수 있습니다.


## 라이브러리 사용

또한 특정 라이브러리를 사용한다면 그 라이브러리의 핵심 가치와 사용 방법을 잘지키는 것도 중요하니 이 부분도 언제나 꼼꼼히 확인해보세요.

특히 `Redux` 같은 라이브러리는 앱의 모든 부분에 많은 영향을 줄 수 있습니다.
때문에 사용법을 잘 지키지 않는다면 예상치 못한 에러를 생각보다 쉽게 만나게 될 수 있고 대응하기도 어렵답니다 ㅠㅠ

---

전반적으로 일관성과 네이밍에 많은 시간을 쓰는게 느껴져 뿌듯합니다.
방학 잘 쉬시고 충전된 모습으로 만나요 :)
